### PR TITLE
Fixed `do_kafka_test.sh` script

### PR DESCRIPTION
**Pull request by dmccoystephenson**
## Changes
The `do_kafka_test.sh` was created six years ago, but it was discovered last year that it was not functioning properly due to inconsistencies in container name references and issues with dependent scripts.

The script has been refactored and thoroughly tested to ensure its functionality when executed in WSL.

## Additional Changes
- The dockerfile.testing file has been added to the project for testing purposes. This includes a python installation.
- The majority of the code has been moved into methods in `do_kafka_test.sh`
- Color has been added to the `do_kafka_test.sh`, `do_bsm_test.sh` and `do_tim_test.sh` scripts.
- The `standalone.sh` and `standalone-multi.sh` scripts have been refactored.
- Descriptions as comments have been added to a number of scripts.
- The DOCKER_HOST_IP environment variable is now required to be set for the `do_kafka_test.sh` script to run.
- The `docker-compose-kafka.yml` file has been added for testing purposes.

## Testing
These changes have been verified to be working when the script is run locally in WSL.

### CODE REVIEW

**Review by drewjj — APPROVED** [review]
This looks good to me

File: Dockerfile-nsv [hunk 1]
```diff
@@ -15,6 +15,9 @@ RUN apt install -y cmake
 # Install librdkafka.
 RUN apt-get install -y libsasl2-dev libsasl2-modules libssl-dev librdkafka-dev
 
+# install python for testing
```
Comment (payneBrandon) [inline Dockerfile-nsv:18]: With the addition of Dockerfile.testing should we drop the python install from this nsv and the standalone dockerfiles?

File: do_kafka_test.sh [hunk 2]
```diff
@@ -1,71 +1,150 @@
 #!/bin/bash
```
Comment (payneBrandon) [inline do_kafka_test.sh:1]: Should we update the testing.md file to show how to run this and what variables are required? Just trying to run this script after opening in the dev container threw and error of "DOCKER_HOST_IP is not set" which could be avoided with some documentation in that markdown file.

**Review by payneBrandon — CHANGES_REQUESTED** [review]
Looking good, just a couple comments

File: Dockerfile-nsv (hunk 1 above)
Comment (dmccoystephenson) [reply Dockerfile-nsv]: Yup good call! I've removed the python installation from these dockerfiles.

File: do_kafka_test.sh (hunk 2 above)
Comment (dmccoystephenson) [reply do_kafka_test.sh:1]: Documentation has been added to the README!

**Comment by payneBrandon** [comment]
@dmccoystephenson the `do_kafka_test.sh` script seems to be failing on all the tests for me

**Comment by dmccoystephenson** [comment]
> Tests currently failing - maybe I missed a step somewhere?

The tests fail sometimes when data cannot be verified to be consumed from the kafka cluster. I have found that the tests are more likely to pass after a fresh restart of the host computer. It's possible that some background processing with docker is causing the issue (which would explain why a restart would help).

**Comment by dmccoystephenson** [comment]
Tests pass for me. I am running `do_kafka_test.sh` in WSL with no other active containers present at the time of execution.

I will continue to experiment until the tests fail and then investigate further.

**Comment by dmccoystephenson** [comment]
Compilation was failing when using Ubuntu 20.04 as the base image. After switching back to using Ubuntu 18.04, the tests are passing again.

**Comment by dmccoystephenson** [comment]
These changes have been re-deployed to our dev cluster and data has been verified to still be flowing.

**Review by drewjj — CHANGES_REQUESTED** [review]
Looks like the tests are not able to consume any of the produced messages. Not sure if this is a timing issue or something else. Definitely needs some more investigation.

**Comment by dmccoystephenson** [comment]
I suspect that the `do_kafka_test.sh` script was failing sometimes because data got produced before the PPM was ready to consume it. I have modified the standalone test scripts to wait until the PPM logs reach 100 lines instead of waiting a fixed 20 seconds for PPM readiness.

Please let me know if you run into more issues @drewjj

**Review by drewjj — CHANGES_REQUESTED** [review]
I am still seeing my tests fail. I am running it in Ubuntu, I have the proper IP information setup due to the script successfully creating the messages but the test seems to fail at reading them for its test case. Unsure what to do here.
image

**Review by drewjj — APPROVED** [review]
The test script works great now. I hadn't fully updated my local repository before.

**Review by dmccoystephenson — COMMENTED** [review]
Self-review

File: README.md [hunk 3]
```diff
@@ -210,6 +210,32 @@ If the RPM_DEBUG environment variable is set to true, debug messages will be log
...
+- test_in.py
+- test_out.py
+
+The DOCKER_HOST_IP environment variable must be set to the IP address of the host machine. This is required for the script to function properly. This can be set by using the following command:
+
+```
+export DOCKER_HOST_IP=$(ifconfig | zgrep -m 1 -oP '(?<=inet\s)\d+(\.\d+){3}')
+```
+
+WSL will sometimes hang indefinitely while script waits for kafka to create topics. When this happens, running `wsl --shutdown` in a windows command prompt and restarting the docker services is recommended.
```
[the remark this answers is not in the document]
Comment (dmccoystephenson) [reply README.md]: Modified this note.

File: do_kafka_test.sh [hunk 4]
```diff
@@ -1,71 +1,159 @@
...
-        break   
-    fi
-
-    sleep 1
-done
+CYAN='\033[0;36m'
+YELLOW='\033[1;33m'
+NC='\033[0m' # No Color
 
+KAFKA_CONTAINER_NAME=jpo-cvdp-kafka-1
```
[the remark this answers is not in the document]
Comment (dmccoystephenson) [reply do_kafka_test.sh]: This variable should be correct regardless of the project directory's name now.

**Review by payneBrandon — APPROVED** [review]

**Pull request merged**